### PR TITLE
Add HAProxy Plugin

### PR DESCRIPTION
**Pull request by ericwholt**
This will add new plugin to parse HAProxy from file. You can select between observiq, http, and tcp log formats. See below for references to the log formats.

### Default Log Format
https://www.haproxy.com/documentation/hapee/latest/onepage/#8.2.1

Example Output:
``` JSON
{
  "timestamp": "2021-09-19T15:49:00-04:00",
  "severity": 0,
  "labels": {
    "file_name": "json.log",
    "log_type": "haproxy.observiq",
    "plugin_id": "haproxy"
  },
  "record": {
    "host": "oiq-int-haproxy",
    "message": "Proxy apache started.",
    "pid": "15202",
    "process_name": "haproxy"
  }
}
```

### TCP Log Format
https://www.haproxy.com/documentation/hapee/latest/onepage/#8.2.2
```
log-format "%ci:%cp [%t] %ft %b/%s %Tw/%Tc/%Tt %B %ts %ac/%fc/%bc/%sc/%rc %sq/%bq"
```
Example Output:
``` JSON
{
  "timestamp": "2021-09-19T15:49:00-04:00",
  "severity": 0,
  "labels": {
    "file_name": "tcp.log",
    "log_type": "haproxy.tcp",
    "plugin_id": "haproxy"
  },
  "record": {
    "accept_date": "06/Feb/2009:12:12:51.443",
    "backend_concurrent_connections": "0",
    "backend_name": "bck",
    "backend_queue": "0",
    "bytes_read": "212",
    "client_ip": "10.0.1.2",
    "client_port": "33313",
    "frontend_concurrent_connections": "0",
    "frontend_name": "fnt",
    "host": "oiq-int-haproxy",
    "pid": "14387",
    "process_concurrent_connections": "0",
    "process_name": "haproxy",
    "queue_wait_time": "0",
    "server_concurrent_connections": "0",
    "server_connection_retries": "3",
    "server_name": "srv1",
    "server_queue": "0",
    "server_response_time": "0",
    "session_duration": "5007",
    "termination_state": "--"
  }
}
```

### Http Log Format
https://www.haproxy.com/documentation/hapee/latest/onepage/#8.2.3

```
log-format "%ci:%cp [%tr] %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}r"
```
Example Output:
``` JSON
{
  "timestamp": "2021-09-19T15:49:00-04:00",
  "severity": 60,
  "severity_text": "503",
  "labels": {
    "file_name": "http.log",
    "log_type": "haproxy.http",
    "plugin_id": "haproxy"
  },
  "record": {
    "accept_date": "16/Sep/2021:16:32:49.352",
    "backend_concurrent_connections": "0",
    "backend_name": "http",
    "backend_queue": "0",
    "bytes_read": "213",
    "captured_request_cookie": "-",
    "captured_response_cookie": "-",
    "client_ip": "127.0.0.1",
    "client_port": "40140",
    "client_request_active_time": "0",
    "client_request_send_time": "-1",
    "frontend_concurrent_connections": "1",
    "frontend_name": "http",
    "host": "oiq-int-haproxy",
    "method": "GET",
    "path": "/",
    "pid": "15165",
    "process_concurrent_connections": "1",
    "process_name": "haproxy",
    "protocol": "HTTP",
    "protocol_version": "1.1",
    "queue_wait_time": "-1",
    "server_concurrent_connections": "0",
    "server_connection_retries": "0",
    "server_name": "<NOSRV>",
    "server_queue": "0",
    "server_response_send_time": "-1",
    "server_response_time": "-1",
    "status": "503",
    "termination_state": "SC--"
  }
}
```

### Observiq Log Format

Will update with log-format

Example Output:
``` JSON
{
  "timestamp": "2021-09-18T15:39:01-04:00",
  "severity": 50,
  "severity_text": "404",
  "labels": {
    "file_name": "json.log",
    "log_type": "haproxy.observiq",
    "plugin_id": "haproxy"
  },
  "record": {
    "frontend_addr": "10.33.104.161",
    "frontend_port": 8080,
    "frontend_ssl_ciphers": "-",
    "frontend_ssl_version": "-",
    "haproxy_backend_concurrent_connections": 0,
    "haproxy_backend_name": "apache",
    "haproxy_backend_queue": 0,
    "haproxy_client_request_send_time": 0,
    "haproxy_frontend_concurrent_connections": 1,
    "haproxy_frontend_name": "http",
    "haproxy_frontend_type": "http",
    "haproxy_process_concurrent_connections": 1,
    "haproxy_queue_wait_time": 0,
    "haproxy_server_concurrent_connections": 1,
    "haproxy_server_connection_retries": 0,
    "haproxy_server_name": "apache",
    "haproxy_server_queue": 0,
    "haproxy_server_response_send_time": 0,
    "haproxy_server_wait_time": 0,
    "host": "10.33.104.161:8080",
    "pid": 15705,
    "process_name": "haproxy",
    "referer": "-",
    "remote_addr": "10.33.104.161",
    "remote_port": 55498,
    "request_http_version": "HTTP/1.1",
    "request_method": "GET",
    "request_size": 136,
    "request_termination_state": "----",
    "request_uri": "/v1/api/app?user=dev",
    "response_size": 438,
    "response_time": 1,
    "session_duration": 1,
    "status": 404
  }
}
```

### CODE REVIEW

**Review by jsirianni — CHANGES_REQUESTED** [review]
Couple things:
- I think TCP logs should default to INFO level logging
- Handle error logs

I think we need to handle error logs, I will post a big example in the PR comments
```json
{
  "timestamp": "2021-09-20T09:41:34-04:00",
  "severity": 0,
  "labels": {
    "file_name": "all.log",
    "log_type": "haproxy.observiq",
    "plugin_id": "haproxy"
  },
  "record": {
    "host": "oiq-int-haproxy",
    "message": "[WARNING] 262/094134 (5149) : parsing [/etc/haproxy/haproxy.cfg:26] : 'option httplog' not usable with frontend 'tcp' (needs 'mode http'). Falling back to 'option tcplog'.",
    "pid": "5149",
    "process_name": "haproxy"
  }
}
```
```
Sep 20 09:41:34 oiq-int-haproxy haproxy[5116]: [WARNING] 262/094045 (5116) : Exiting Master process...
Sep 20 09:41:34 oiq-int-haproxy haproxy[5116]: [ALERT] 262/094045 (5116) : Current worker 5118 exited with code 143
Sep 20 09:41:34 oiq-int-haproxy haproxy[5116]: [WARNING] 262/094045 (5116) : All workers exited. Exiting... (143)
Sep 20 09:41:34 oiq-int-haproxy haproxy[5149]: [WARNING] 262/094134 (5149) : parsing [/etc/haproxy/haproxy.cfg:26] : 'option httplog' not usable with frontend 'tcp' (needs 'mode http'). Falling back to 'option tcplog'.
Sep 20 09:41:34 oiq-int-haproxy haproxy[5149]: [ALERT] 232/194354 (199) : parsing [/etc/haproxy/haproxy.cfg:13] : unknown keyword 'bind' in 'global' section
```

I think the fields could be
- severity
- message (hard to parse the rest)

**Comment by jsirianni** [comment]
Expanding on my feedback: Lets remove `log_format` and check if http / tcp / json. The reason for this is because of how easy it is to configure haproxy to use all 4 formats (two default, two json)
```
frontend tcp
    bind 0.0.0.0:8081
    mode tcp
    default_backend apache

frontend tcp-json
    bind 0.0.0.0:8083
    mode tcp
    default_backend apache
    log-format '{"pid":%pid,"haproxy_frontend_type":"tcp","haproxy_process_concurrent_connections":%ac,"haproxy_frontend_concurrent_connections":%fc,"haproxy_backend_concurrent_connections":%bc,"haproxy_server_concurrent_connections":%sc,"haproxy_backend_queue":%bq,"haproxy_server_queue":%sq,"haproxy_queue_wait_time":%Tw,"haproxy_server_wait_time":%Tc,"response_time":%Td,"session_duration":%Tt,"request_termination_state":"%tsc","haproxy_server_connection_retries":%rc,"remote_addr":"%ci","remote_port":%cp,"frontend_addr":"%fi","frontend_port":%fp,"frontend_ssl_version":"%sslv","frontend_ssl_ciphers":"%sslc","haproxy_frontend_name":"%f","haproxy_backend_name":"%b","haproxy_server_name":"%s","response_size":%B,"request_size":%U}'

frontend http
    bind 0.0.0.0:8080
    mode http
    default_backend apache
    # When in HTTP mode
    http-request capture req.hdr(Host) len 1000
    http-request capture req.hdr(Referer) len 1000

frontend http-json
    bind 0.0.0.0:8082
    mode http
    default_backend apache
    # When in HTTP mode
    http-request capture req.hdr(Host) len 1000
    http-request capture req.hdr(Referer) len 1000
    log-format '{"pid":%pid,"haproxy_frontend_type":"http","haproxy_process_concurrent_connections":%ac,"haproxy_frontend_concurrent_connections":%fc,"haproxy_backend_concurrent_connections":%bc,"haproxy_server_concurrent_connections":%sc,"haproxy_backend_queue":%bq,"haproxy_server_queue":%sq,"haproxy_client_request_send_time":%Tq,"haproxy_queue_wait_time":%Tw,"haproxy_server_wait_time":%Tc,"haproxy_server_response_send_time":%Tr,"response_time":%Td,"session_duration":%Tt,"request_termination_state":"%tsc","haproxy_server_connection_retries":%rc,"remote_addr":"%ci","remote_port":%cp,"frontend_addr":"%fi","frontend_port":%fp,"frontend_ssl_version":"%sslv","frontend_ssl_ciphers":"%sslc","request_method":"%HM","request_uri":"%[capture.req.uri,json(utf8s)]","request_http_version":"%HV","host":"%[capture.req.hdr(0)]","referer":"%[capture.req.hdr(1),json(utf8s)]","haproxy_frontend_name":"%f","haproxy_backend_name":"%b","haproxy_server_name":"%s","status":%ST,"response_size":%B,"request_size":%U}'


backend apache
    mode http
    server apache localhost:80
```

**Comment by jsirianni** [comment]
rename regex capture group fields:
```
server_response_time --> response_time
server_response_send_time --> response_send_time
```

**Comment by jsirianni** [comment]
URI Parser
1. rename `path` to `uri` (its called `uri` in the json format, so we should be consistent with our regex capture group names)
2. parse `uri` with `uri_parser`

Ideally something like this:
```
uri: /app?user=admin
```
will end up like this:
```yaml
path: /app
query:
  user
    - admin
```

So we will end up with a `path` field still, but it wont include the query parameters, which is more correct.

**Comment by jsirianni** [comment]
default format:
```
protocol: HTTP
protocol_version: 1.1
```

http_version parsing. JSON format has
```
http_version: HTTP/1.1
```

We need to unify these, probably by parsing json format into `protocol` and `protocol_version` and removing `http_version`

**Comment by jsirianni** [comment]
I replaced the log lobrary json.log with all.log, which includes json, tcp, http in a combined log file as well as errors.

I also updated the json log format. This new format has field name parity with the names you decided on, or with the feedback I have already provided with respect to field name changes.

tcp
```
'{"pid":%pid,"frontend_type":"tcp","process_concurrent_connections":%ac,"frontend_concurrent_connections":%fc,"backend_concurrent_connections":%bc,"server_concurrent_connections":%sc,"backend_queue":%bq,"server_queue":%sq,"queue_wait_time":%Tw,"server_wait_time":%Tc,"response_time":%Td,"session_duration":%Tt,"termination_state":"%tsc","server_connection_retries":%rc,"client_ip":"%ci","client_port":%cp,"frontend_ip":"%fi","frontend_port":%fp,"frontend_ssl_version":"%sslv","frontend_ssl_ciphers":"%sslc","frontend_name":"%f","backend_name":"%b","server_name":"%s","bytes_read":%B,"bytes_uploaded":%U}'
```
http
```
'{"pid":%pid,"frontend_type":"http","process_concurrent_connections":%ac,"frontend_concurrent_connections":%fc,"backend_concurrent_connections":%bc,"server_concurrent_connections":%sc,"backend_queue":%bq,"server_queue":%sq,"client_request_send_time":%Tq,"queue_wait_time":%Tw,"server_wait_time":%Tc,"server_response_send_time":%Tr,"response_time":%Td,"session_duration":%Tt,"termination_state":"%tsc","server_connection_retries":%rc,"client_ip":"%ci","client_port":%cp,"frontend_ip":"%fi","frontend_port":%fp,"frontend_ssl_version":"%sslv","frontend_ssl_ciphers":"%sslc","method":"%HM","uri":"%[capture.req.uri,json(utf8s)]","http_version":"%HV","referer":"%[capture.req.hdr(1),json(utf8s)]","frontend_name":"%f","backend_name":"%b","server_name":"%s","status":%ST,"bytes_read":%B,"bytes_uploaded":%U}'
```

**Comment by jsirianni** [comment]
rename server_connection_retries --> retries

**Comment by jsirianni** [comment]
Circling back to removing log_format, I think it should be optional. If not set, then we should detect the format. If set to default, then detect tcp / http. If set to observiq, use json.

This way, if the user is using observiq format, they get the performance boost by skipping the auto detection.

**Comment by jsirianni** [comment]
Promote the following fields to labels / resources. I think we can skip checking if the field exists because regex parser and json parser will always return these fields even if they are empty.

labels
```
frontend_ssl_version
frontend_ssl_ciphers
method
http_version
protocol
protocol_version
referer
```
resources
```
frontend_type
frontend_ip
frontend_port
path
frontend_name
backend_name
server_name
```

**Comment by jsirianni** [comment]
set min stanza version 1.2.6

File: plugins/haproxy.yaml [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    label: Start At
+    description: Start reading file from 'beginning' or 'end'
+    type: enum
+    valid_values:
+     - beginning
+     - end
+    default: end
+# Set Defaults
+#{{$start_at := default "end" .start_at}}
+# {{$log_format := default "default" .log_format}}
```
[the remark this answers is not in the document]
Comment (ericwholt) [reply plugins/haproxy.yaml:32]: Fixed spacing in next commit.

File: plugins/haproxy.yaml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+min_stanza_version: 0.14.0
+parameters:
+  - name: file_log_path
+    label: File Path
+    description: Specify a single path or multiple paths to read one or many files. You may also use a wildcard (*) to read multiple files within a directory.
+    type: strings
+    required: true
+  - name: log_format
+    label: Log Format
+    description: When choosing the 'default' option, the agent will expect and parse logs in a format of HTTP or TCP as well as any log entries that matches the default logging configuration. HAProxy uses default logging format when no specific option is set. When choosing the 'observIQ' option, the agent will expect and parse logs in an optimized JSON format that adheres to the observIQ specification, requiring an update to the Log-Format for each mode. See the HAProxy source page for more information.
```
[the remark this answers is not in the document]
Comment (ericwholt) [reply plugins/haproxy.yaml:15]: Left default as we added a default option due to next suggestion

File: plugins/haproxy.yaml [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    type: enum
+    valid_values:
+      - http
+      - tcp
+      - observiq
+    default: http
```
[the remark this answers is not in the document]
Comment (ericwholt) [reply plugins/haproxy.yaml:21]: Done

**Review by schmikei — APPROVED** [review]
